obliterator

Cokemonkey11 opened the pull request:
- [x] Untested because of https://github.com/wurstscript/WurstScript/issues/1034

Cokemonkey11 committed:
obliterator

Cokemonkey11 commented on wurst.build:
```diff
@@ -1,5 +1,5 @@
 ---
-projectName: Gods' Arena
+projectName: Gods Arena
```
WurstSetup wouldn't let me update project with old value

Frotty replied:
don't change project name. We are having no problems using grill.

Cokemonkey11 commented on wurst.build:
```diff
@@ -17,8 +17,14 @@ buildMapData:
-        \ (hotkey - F).\n\n You will be revived at the end of each round, but not\
-        \ with full HP and MP.\n\n You will restore a bit of HP and MP at the end\
-        \ of each round if you survived."
+      text: |-
```
WurstSetup did this automatically

Frotty replied:
And? Probably using old version, don't commit this stuff.

Cokemonkey11 commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,56 @@
+
+      if curr_bonus != wants_bonus
+        u.addStr(wants_bonus - curr_bonus)
+        print(wants_bonus - curr_bonus)
```
Need to delete after testing

Frotty commented on wurst/foes/creeps/CreepAggro.wurst:
```diff
@@ -17,7 +17,13 @@ class AggroEntry
 constant aggroMap = new HashMap<unit, LinkedList<AggroEntry>>()
 
-public function unit.addAggro(unit target, int amount)
+interface AggroMergeFn<T>
```
Why this overcomplication? Not a fan of this

Cokemonkey11 replied:
@Frotty can you explain your objection more? I do not see a more elegant way to provide both `add` and `set` function in your system. This is just [merge](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#merge-K-V-java.util.function.BiFunction-)

Frotty replied:
I guess the overuse of closures. On second thought, impl is okay, but more important for lowering aggro andthan "setting" it i think.

Cokemonkey11 replied:
I was hoping that since there is constant closure for add/set cases, then the performance impact would be acceptable

Thinking about this more, and about your comment regarding 200 being too low for late game, i wonder if this should actually be yet a third merge function, `max`, which uses max of provided and current values. I would be interested in your opinion.

Frotty commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,56 @@
+  return "custom_" + this.toRawCode()
+
+init
+  doPeriodically(1.) _ ->
```
Why is this running globally like this? Heroes can only exist once.

Overkane replied:
Probably could use damage event instead?

Cokemonkey11 replied:
@Frotty I thought to make it MUI but yeah unnecessary, also doesn't work. Fixing now.

@Overkane I don't think so - the idea of the ability is to passively aggro all enemies nearby

Frotty replied:
You didn't answer why this runs periodically. Are you evaluating the "buff" here? then you should use a proper managed visual buff using frentity, instead of your own loop timer. should be better anyway.

Cokemonkey11 replied:
The reason why this runs periodically is to:

- add or remove bunus strength as appropriate
- set/max aggro of all adjacent enemies

I'm not familiar with frentity visual buff, but i can check that if you think it would work here.

I'm interested to know what your objection to the timer is and why frentity would be better in that case. Perhaps i can find this out myself

Cokemonkey11 replied:
@Frotty I had a quick search in frentity repo but I don't see anything about visual buff

Overkane commented:
hm, did you test the spell? It doesn't work for me

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,56 @@
+    Targettype.PASSIVE,
+    "Passively grants bonus strength and nearby creep aggro when the Demon Lord has less than 50% HP."
+  )
+  new AbilityDefinitionArchMageBrillianceAura(OBLITERATOR_ID)
```
Tooltip doesn't work and it's preferably to use PassiveAbilityPreset from the same named package for custom passives

Cokemonkey11 commented:
@Frotty @Overkane I cannot really fix/test this because of https://github.com/wurstscript/WurstScript/issues/1034

I converted to draft for now

Cokemonkey11 committed:
cr

Cokemonkey11 commented on wurst/heroes/heroes/DemonLord/HowlOfTerror.wurst:
```diff
@@ -2,25 +2,37 @@ package HowlOfTerror
 RealLevelClosure DAMAGE_DECREASE = lvl -> 0.25 + 0.1 * (lvl-1)
 RealLevelClosure DURATION = lvl -> 10
-IntLevelClosure STR_BONUS = lvl -> (0.5 * (lvl - 1)).floor() + 1 // 1, 1, 2, 2, 3
+IntLevelClosure STR_BONUS = _ -> 1
```
This was overpowered, could nearly solo first boss after scaling

Overkane replied:
The problem that solo you get all the kills and can get large amount of strength
But when you play with other 7 players that will be hard to get a kill.
But seems fine

Cokemonkey11 replied:
In a follow-up PR i would likely to add some hero effectiveness stats, to make this more objective/observable

Cokemonkey11 commented on wurst/heroes/heroes/DemonLord/HowlOfTerror.wurst:
```diff
@@ -2,25 +2,37 @@ package HowlOfTerror
-      killer.addStr(STR_BONUS.run(level))
-      killer.addHP(HP_RECOVERY.run(level))
-      flashEffect(Abilities.deathCoilMissile, killer.getPos())
+  DamageEvent.addListener(5) ->
```
Old code was not working - EVENT_UNIT_DEATH event fires after buffs are removed from unit

Overkane replied:
yea, I fixed that and also added aggro to the howl
did you merge this package when you merged with master?

Cokemonkey11 replied:
Yes merged in 1f1097d

Overkane replied:
![image](https://user-images.githubusercontent.com/29690395/138141955-af26cd3d-a972-446a-99f5-51b9368df50b.png)
but why is this part missing? I added aggro and fixed that bug and frotty fixed addHp to custom heal system

Cokemonkey11 replied:
Don't know, I will just revert, my change shouldn't be in this PR anyway

Cokemonkey11 committed:
Merge branch 'master' into cokemonkey11/obliterator

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,79 @@
+    if this.fx != null
+      this.fx.destr()
+
+function int.toCustomUnitTypeString() returns string
```
not used?

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,79 @@
+      if wants
+        forUnitsInRange(u.getPos(), CREEP_AGGRO_RANGE.run(1)) fog ->
+          if fog.isEnemyOf(u)
+            u.setAggro(fog, CREEP_AGGRO.run(1))
```
why do you set aggro, but not add?

Cokemonkey11 replied:
Because adding would ramp up to infinite aggro on timer system

But this should be `max(real, real)` - not add nor set. Fixing.

Overkane reviewed (changes requested):
- map name should stay, since it doesn't refer to the commit. I can change it later by another commit.
- howl of terror package should be merged again and consider new fixes to it. It won't be changed on master for some time I believe.
- can't say much about CreepAggro.wurst additions, if it works, then it fine I guess.


Obliterator spell.
1. It doesn't work well.
I got below 50% of hp.
I got 50 str and cuz of it health went to ~70% of hp.
When I got below 50% of hp again, 50 str stayed, but I healed a bit to like 55% of hp.
And then it repeats, so you just can't die, cuz you heal everytime you get below 50% of hp.
2. Anyway, that's 50 str forever? Or when you get above 50% of hp, you get -50 str?
3. OblitiratorData is here to make the spell MUI? But I think you could make it simple, consider that there is can only be the one demon lord at the time.
4. If I not misunderstood, the spell set aggro to 200 for near creeps when hero below 50% of hp?

Frotty commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,79 @@
+import PlayerData
+
+constant IntLevelClosure BONUS_STRENGTH = _ -> 50
+constant RealLevelClosure CREEP_AGGRO = _ -> 200.
```
way too low. aggro == damage. 200 can be one hit in late game

Frotty commented:
Coke didn't merge, he just took his changesOn Oct 20, 2021 19:29, Stanislav ***@***.***> wrote:
@Overkane commented on this pull request.



In wurst/heroes/heroes/DemonLord/HowlOfTerror.wurst:
>  RealLevelClosure HP_RECOVERY = lvl -> 100.0 + 50 * (lvl-1)
 
 init
-  EventListener.add(EVENT_PLAYER_UNIT_DEATH) ->
-    let killer = EventData.getKillingUnit()
-    if killer.getTypeId() == DEMON_LORD_ID and EventData.getTriggerUnit().hasAbility(BuffIds.howlofTerror)
-      let level = killer.getAbilityLevel(HOWL_OF_TERROR_ID)
-      killer.addStr(STR_BONUS.run(level))
-      killer.addHP(HP_RECOVERY.run(level))
-      flashEffect(Abilities.deathCoilMissile, killer.getPos())
+  DamageEvent.addListener(5) ->


but why is this part missing? I added aggro and fixed that bug and frotty fixed addHp to custom heal system

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

Cokemonkey11 committed:
cr

Cokemonkey11 commented:
> 
> Coke didn't merge, he just took his changesOn Oct 20, 2021 19:29, Stanislav ***@***.***> wrote: @Overkane commented on this pull request. In wurst/heroes/heroes/DemonLord/HowlOfTerror.wurst:
> RealLevelClosure HP_RECOVERY = lvl -> 100.0 + 50 * (lvl-1)
> init - EventListener.add(EVENT_PLAYER_UNIT_DEATH) -> - let killer = EventData.getKillingUnit() - if killer.getTypeId() == DEMON_LORD_ID and EventData.getTriggerUnit().hasAbility(BuffIds.howlofTerror) - let level = killer.getAbilityLevel(HOWL_OF_TERROR_ID) - killer.addStr(STR_BONUS.run(level)) - killer.addHP(HP_RECOVERY.run(level)) - flashEffect(Abilities.deathCoilMissile, killer.getPos()) + DamageEvent.addListener(5) -> but why is this part missing? I added aggro and fixed that bug and frotty fixed addHp to custom heal system —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

nah just botched 3-way diff, didn't read your changes that carefully. reverted my change on this area now

Cokemonkey11 commented:
> map name should stay, since it doesn't refer to the commit. I can change it later by another commit.

✅ 

> howl of terror package should be merged again and consider new fixes to it. It won't be changed on master for some time I believe.

Reverted my changes, it was stupid to bother with merge

> can't say much about CreepAggro.wurst additions, if it works, then it fine I guess.

Yep it works

> Obliterator spell. It doesn't work well. I got below 50% of hp. I got 50 str and cuz of it health went to ~70% of hp. When I got below 50% of hp again, 50 str stayed, but I healed a bit to like 55% of hp. And then it repeats, so you just can't die, cuz you heal everytime you get below 50% of hp.

I tested this carefully now and it seems that there is a problem [here](https://github.com/Overkane/gods-arena/blob/1f1097d2f0e96177d1ed21951681be926b6f86cd/wurst/heroes/heroes/DemonLord/Obliterator.wurst#L46L48). The expected behavior is that (1) remove strength (2) check health (3) re-add strength would achieve two things (check health without bonus, to find out if bonus should still be applied, (2) not change health

In fact, it seems that removing and re-adding strength actually heals a bit, because removing strength is not allowed to reduce hero's HP below 1.0:

![image](https://user-images.githubusercontent.com/1418966/138530820-99e39861-73ef-42c7-bbf3-82c895ea29b7.png)

Added a note to [jassdoc about it](https://github.com/lep/jassdoc/pull/38)

> Anyway, that's 50 str forever? Or when you get above 50% of hp, you get -50 str?

Yes, that's how it works, but it should calculate "above 50% hp" without the bonus strength (linked above)

> OblitiratorData is here to make the spell MUI? But I think you could make it simple, consider that there is can only be the one demon lord at the time.

No, ObliteratorData enables idempotent management of vampiring aura effect. DEMON_LORD_STR_BONUSES is for MUI. I can change this if you want, but I think it doesn't matter.

> If I not misunderstood, the spell set aggro to 200 for near creeps when hero below 50% of hp?

Yes, now increased to 400 and changed from "set" function to "math max" function.

@Overkane based on the unlimited healing notes above, I would have to rethink how to make this work reliably.

Do you prefer if I scrap it and make a new idea, or revisit how to make it work? I think it depends - do you like the design or not?

Overkane commented:
I think `max` function looks strange here.
Let's say we have used lvl 4 firestorm and dealt 525 damage to near creeps. Then a mage hero casts his spell for 600 damage.
Even if demon lord goes below 50% of hp, it will do nothing with aggro, since 400 < 525 and even if passive spell aggro was bigger than aggro from firestorm it still always will be less than 600 for example.
So I don't see, how it fits, maybe could be a think only in early game.

Overall I like that kind of "enrage" mechanic. Seems fitting for demon hero.
For example it could activate if hero gets below 35% of hp. Then he gets bonus 15% of lifesteal and 90 of attack speed for 10 second. To prevent unlimited activation, passive has 50 seconds cd after activation.
Something like that would be nice.

Cokemonkey11 committed:
switch to agi

Cokemonkey11 commented:
@Overkane 

> So I don't see, how it fits, maybe could be a think only in early game.

Yep, that is how it works. The flat bonus also has more relevance in early game than late game. I think it makes the behavior more interesting, personally.

> Something like that would be nice.

Okay, I like your idea too. I thought about it more and realised that this isn't much different than simply switching strength bonus to agi bonus, which doesn't have the "healing problem" that str has. Let me know what you think.

IMO this is good to go.

Frotty commented:
I think you should actually play test the hero and don't just assume stuff works fine, especially in late game.
By visual buff I mean a buff that u can see in the unit status bar, and if u can't find Buff.wurst in frentity I already told you that other heroes in god's arena have buffs which you can check.
The bonus calculation now doesn't make much sense, because agility bonus doesn't affect health.
And still, no rhyme or reason for the doPeriodically. The spell says "when life goes below X, apple, bonus Y. Then I expect a damage listener which applied a buff when the hero is below the threshold and that buff should handle the bonus. Also it is not clear from the tooltip, how the bonuses are removed again. Do u insta lose everything when health goes above again?

Overkane commented:
> @Overkane
> 
> > So I don't see, how it fits, maybe could be a think only in early game.
> 
> Yep, that is how it works. The flat bonus also has more relevance in early game than late game. I think it makes the behavior more interesting, personally.
> 
> > Something like that would be nice.
> 
> Okay, I like your idea too. I thought about it more and realised that this isn't much different than simply switching strength bonus to agi bonus, which doesn't have the "healing problem" that str has. Let me know what you think.
> 
> IMO this is good to go.



> Let me know what you think.

I think that's strange, that str hero gets agi bonus so instead of attributes could switch to something else.

Cokemonkey11 commented:
@Frotty 

> I think you should actually play test the hero and don't just assume stuff works fine, especially in late game.

I have definitely play-tested, but only up to ~first boss. I think it makes more sense to agree general design is interesting before refining specifics and playtesting balance.

I also disagree fundamentally that there is anything in _need_ of balance testing here. At the end of the day, there are much bigger balance concerns and without metrics, every judgment about balance is prone to bias.

> By visual buff I mean a buff that u can see in the unit status bar, and if u can't find Buff.wurst in frentity I already told you that other heroes in god's arena have buffs which you can check.

Nice, thanks, I found it and integrated. It is pretty elegant apart from object hierarchy.

>  The bonus calculation now doesn't make much sense, because agility bonus doesn't affect health.

Thanks, fixed.

>  And still, no rhyme or reason for the doPeriodically. The spell says "when life goes below X, apple, bonus Y. Then I expect a damage listener which applied a buff when the hero is below the threshold and that buff should handle the bonus. Also it is not clear from the tooltip, how the bonuses are removed again. Do u insta lose everything when health goes above again?

Yes, it's meant to lose everything when HP goes back up. But I found a nice way to do this with frentity::Buff 👍 

@Overkane 

> I think that's strange, that str hero gets agi bonus so instead of attributes could switch to something else.

Can you explain this a little more? Do you mean that you feel that it's thematically strange to add agi since it's a str hero, and you prefer to add (attack speed, movespeed, armor) since thematically it seems more neutral?

Cokemonkey11 committed:
use frentity::Buff

Cokemonkey11 commented:
@Frotty @Overkane please note that I ran out of time writing the above, haven't playtested latest version using frentity. I will address any further comments and test again tonight

Overkane commented on wurst/heroes/heroes/DemonLord/PermanentBuff.wurst:
```diff
@@ -0,0 +1,26 @@
+  When applied to a unit that already has the buff, nothing happens. Buff is only removed if the unit dies or if the
+  buff is programatically removed from the unit.
+ */
+public abstract class PermanentBuff extends Buff
```
probably could just use `FixedBuff` with long duration from `PresetBuffs.wurst` (frentity package) instead

Cokemonkey11 replied:
@Frotty I think it's fair if @Overkane doesn't want this merged. Would you accept PermanentBuff contribution to Frentity?

Overkane commented:
> @Overkane
> 
> > I think that's strange, that str hero gets agi bonus so instead of attributes could switch to something else.
> 
> Can you explain this a little more? Do you mean that you feel that it's thematically strange to add agi since it's a str hero, and you prefer to add (attack speed, movespeed, armor) since thematically it seems more neutral?

Yes. With agility, hero would also get an armor bonus. 
That would be fine, if hero would be str and agi user (like gets benefits from both, not including attribute bonuses)
But he isn't.

So I think this ability could be like enrage, which gives bonuses only to attack, cuz that would be strange, if in rage you get defensive bonuses. That's more fitting to current hero, considering he is a demon, so he also can take a good use of his splash.

Cokemonkey11 committed:
use attack speed bonus instead of agi

Cokemonkey11 commented:
@Overkane okay, switched from agi bonus to two other bonuses:

- attack speed bonus
- huge bonus damage against units with less than 5% HP

Also tested and everything looks good to me.

Cokemonkey11 committed:
inline permanentbuff

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,83 @@
+
+  override function update()
+    super.update()
+    if (this.target.getHP() / this.target.getMaxHP()) > BONUS_HEALTH_THRESHOLD.run(1)
```
could use here your function `unit.hpRatio()`?

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,83 @@
+
+
+init
+  DamageEvent.addListener(1) ->
```
could omit `1` here?

Cokemonkey11 replied:
I want to be sure that this listener resolves _after_ any shield listeners modify event damage

Overkane replied:
you can check `DamageSystem.wurst`. Damage modifications are done in unreducedListeners, so reducedListener has final damage.
There is a problem though, that's I implemented lifesteal in `addListener()` (no argument) so I guess this bonus damage won't be included for lifesteal, but I think I'll adjust it by myself later.

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,83 @@
+constant RealLevelClosure SLAY_BONUS_DAMAGE_MULTIPLIER = _ -> 10.
+
+
+public class ObliteratorBuff extends NormalBuff
```
I would like for this buff to be in separate package, like all other buffs

Overkane commented on wurst/heroes/heroes/DemonLord/Obliterator.wurst:
```diff
@@ -0,0 +1,83 @@
+    "Passively " + effect_description + " when the Demon Lord has low HP."
+  )
+  new PassiveAbilityPreset(OBLITERATOR_ID, 1, tooltip)
+  ..presetTargetsAllowed(_ -> "none")
```
not actually need this?

Overkane reviewed (commented):
Seems nice. You could make last adjustments I mentioned and I think we are good.
@Frotty what do you think?

Cokemonkey11 committed:
cr

Overkane commented:
seems nice, thank you!
